File: python/day03/part1.py
#!/usr/bin/env python3
import string
import time


def parse(input_path):
    with open(input_path, 'r') as f:
        return [line.strip() for line in f.readlines()]


def priority(c):
    return string.ascii_letters.index(c) + 1


def solve_part1(bags):
    res = 0
    for bag in bags:
        for item_type in bag[:len(bag)//2]:
            if item_type in bag[len(bag)//2:]:
                res += priority(item_type)
                break
    return res


if __name__ == "__main__":
    start = time.time()
    parsed = parse("data.txt")
    result = solve_part1(parsed)
    print(f'Part 1 : {result}  ({round(1000 * (time.time() - start))} ms)')

File: c++/day03/part1.cpp
#include "part1.h"

using namespace std;


namespace Day03 {

    vector<Bag> parse(const string& filename) {
        vector<Bag> parsed;
        for (const string& line : getFileLines(filename)) {
            const string half1 = line.substr(0, line.size() / 2);
            const string half2 = line.substr(line.size() / 2);
            set<char> list1 = { half1.begin(), half1.end() };
            set<char> list2 = { half2.begin(), half2.end() };
            parsed.emplace_back(list1, list2);
        }
        return parsed;
    }

    namespace Part1 {
        int solve(const vector<Bag>& bags) {
            int res = 0;
            for (const auto& [half1, half2] : bags) {
                for (const char item : half1) {
                    if (half2.find(item) != half2.end()) {
                        res += (item > 'Z') ? item - 'a' + 1 : item - 'A' + 27;
                        break;
                    }
                }
            }
            return res;
        }
    }
}
